File: hw/pazmino/u3/HW37StrategyExercise/HW38StrategySortApp/src/main/java/ec/edu/espe/hw38strategysortapp/model/QuickSort.java
package ec.edu.espe.hw38strategysortapp.model;

/**
 *
 * @author Brandon Pazmino
 */
public class QuickSort implements SortingStrategy {

    @Override
    public int[] sort(int[] data) {
        System.out.println("Using QuickSort...");
        quickSort(data, 0, data.length - 1);
        return data;

    }
    private void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            int pivotIndex = partition(arr, low, high);
            quickSort(arr, low, pivotIndex - 1);
            quickSort(arr, pivotIndex + 1, high);
        }
    }
    
    private int partition(int[] arr, int low, int high) {
        int pivot = arr[high];
        int i = (low - 1);
        for (int j = low; j < high; j++) {
            if (arr[j] <= pivot) {
                i++;
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;
        return i + 1;
    }

}

File: EXAM/bohorquez/u3/FinalExam/espe.edu.ec.model/QuickSort.js
import SortStrategy from "./SortStrategy.js";

class QuickSort extends SortStrategy {
    sort(array) {
        let arr = [...array];
        return this.quickSort(arr, 0, arr.length - 1);
    }

    quickSort(arr, low, high) {
        if (low < high) {
            let pi = this.partition(arr, low, high);
            this.quickSort(arr, low, pi - 1);
            this.quickSort(arr, pi + 1, high);
        }
        return arr;
    }

    partition(arr, low, high) {
        let pivot = arr[high];
        let i = low - 1;
        for (let j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;
                [arr[i], arr[j]] = [arr[j], arr[i]];
            }
        }
        [arr[i + 1], arr[high]] = [arr[high], arr[i + 1]];
        return i + 1;
    }
}

export default QuickSort;
